### PR TITLE
Add pre-parse router for CLI subcommands

### LINKED ISSUE #39: refactor: add optional subcommands to CLI while preserving backward compatibility
## ⚠️ HIGH RISK — PHASE B GATE

This is the single most critical ticket in Phase B. Every subsequent ticket depends on it. A regression here breaks all 400 existing tests and the entire v1.0.0 release.

**Primary risks:**
- Clap ambiguity: `skim test.ts` (positional FILE) vs `skim test cargo` (subcommand). How does clap disambiguate?
- Existing flags like `--mode`, `--tokens`, `--max-lines` must continue working for file operations
- The `skim init` filename edge case — `init` could be a file or a subcommand

**De-risk strategy:**
1. **Clap's `#[command(subcommand_required = false)]`**: Makes subcommands optional, so bare `skim file.ts` still works as positional arg.
2. **Exhaustive disambiguation test matrix**: Test these exact cases before merge:
   - `skim test.ts` → file read (not subcommand)
   - `skim test cargo` → subcommand
   - `skim init` → subcommand (not a file called "init")
   - `skim init.py` → file read
   - `skim src/test.rs --mode=signatures` → file read with flag
   - `skim --help` → combined help showing both file usage and subcommands
3. **All 400 existing tests must pass with zero modifications**: If any test needs changing, the refactor is wrong.
4. **Ship B0-1 alone in its own PR**: Do not bundle with any other work. Review carefully before merging.

---

Add `Option<Commands>` enum to clap so `skim file.ts` still works AND `skim init`, `skim test`, `skim git`, `skim build`, `skim rewrite` work as subcommands. Create `crates/rskim/src/cmd/` module directory.

GRANITE reference: GRANITE uses flat subcommands (`granite git status`). We use optional subcommands to maintain backward compat.

Architecture note: The command dispatch in `cmd/mod.rs` should use a trait-based parser pattern (`OutputParser` trait) that will later be implementable via TOML filter definitions (Phase C).

Key files: `crates/rskim/src/main.rs` (refactor Args), new `crates/rskim/src/cmd/mod.rs`

Acceptance:
- All 400 existing tests pass
- `skim file.ts` unchanged
- `skim init --help` works

Size: M
Depends on: nothing
Blocks: B0-2, B0-3, B0-4, all subsequent phases

### PR DESCRIPTION
**Pull request by dean0x**
## Summary
- Adds pre-parse routing layer that inspects `std::env::args()` before clap to disambiguate file operations from subcommands
- All 5 planned subcommands (`init`, `test`, `rewrite`, `git`, `build`) are stubs printing "not yet implemented"
- Existing `main()` body extracted byte-for-byte into `run_file_operation()` — zero logic changes
- 19 new disambiguation tests, all 471 tests pass

## Design

Uses a **pre-parse router** (same pattern as `cargo`, `docker`, `git`) instead of clap's native subcommand system, which cannot disambiguate `skim test.ts` (file) from `skim test` (subcommand).

### Disambiguation rules (priority-ordered, first match wins)

| Condition | Route |
|-----------|-------|
| No positional arg | FileOperation |
| `--` before first positional | FileOperation |
| Contains `.` | FileOperation |
| Contains `/` or `\` | FileOperation |
| Is `-` | FileOperation |
| Contains `*`, `?`, `[` | FileOperation |
| Known subcommand + no file on disk | Subcommand |
| Everything else | FileOperation |

### Exit code semantics
- `--help`: exit 0, prints stub help to stdout
- Stub invocation: exit 1, prints "not yet implemented" to stderr

## Files changed
- `crates/rskim/src/main.rs` — routing layer added, `main()` body extracted
- `crates/rskim/src/cmd/mod.rs` — NEW: subcommand infrastructure
- `crates/rskim/tests/cli_subcommand.rs` — NEW: 19 disambiguation tests

## Known trade-offs
- Shell completions (#63) will need manual generation since subcommands live outside clap's `Args` struct
- `is_flag_with_value()` must stay in sync with `Args` flags (documented with SYNC NOTE comment)
- `std::process::exit()` on subcommand path skips Drop cleanup (safe for stubs, revisit when real handlers ship)

Closes #39

### CODE REVIEW

File: crates/rskim/tests/cli_subcommand.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            .stderr(predicate::str::contains("not yet implemented"));
+    }
+}
+
+// ============================================================================
+// File-named-as-subcommand precedence
+// ============================================================================
+
+#[test]
+fn test_existing_file_named_as_subcommand_routes_to_file_operation() {
```
Comment (dean0x) [inline crates/rskim/tests/cli_subcommand.rs]: ## CRITICAL TEST GAP: File-on-disk check never exercised

This test passes the **full absolute path** to the file (e.g., `/tmp/xyz/init`). Because the full path contains `/`, the pre-parse router short-circuits at line 117 ("contains '/'") and never reaches the `path.exists()` check at line 128-133. This means the critical backward-compatibility guarantee -- that a bare subcommand name matching an on-disk file in CWD is treated as a file -- is **not actually tested**.

**Fix**: Use `.current_dir()` and pass the bare name `"init"` so the router reaches the `path.exists()` branch:

```rust
#[test]
fn test_bare_subcommand_name_file_in_cwd_routes_to_file_operation() {
    let dir = TempDir::new().unwrap();
    let file = dir.path().join("init");
    fs::write(&file, "fn setup() {}").unwrap();

    // Bare name "init" without path separator → reaches path.exists() check
    Command::cargo_bin("skim")
        .unwrap()
        .current_dir(dir.path())
        .arg("init")
        .arg("-l")
        .arg("rust")
        .assert()
        .success()
        .stdout(predicate::str::contains("fn setup"));
}
```

Confidence: 90% (HIGH)

**Comment by dean0x** [comment]
## Code Review Summary

### Inline Comments Created (≥80% confidence)
- ✅ `std::process::exit()` bypasses destructors at line 396 (82% confidence)
- ✅ `is_flag_with_value()` sync risk at line 41 (85% confidence)
- ✅ `dispatch()` dual error channel at cmd/mod.rs line 56 (85% confidence)
- ✅ Test gap: CWD-relative bare file named as subcommand at line 191 (90% confidence)
- ✅ Test gap: CWD-relative bare directory named as subcommand at line 208 (90% confidence)

### Lower-Confidence Suggestions (60-79%, added to summary only)

**Code Quality Suggestions:**

1. **Error message style inconsistency** (Confidence: 83%) - The error message at `cmd/mod.rs:37` (`"unknown subcommand: {subcommand}"`) uses lowercase and is terse. Consider matching the multi-line pattern used elsewhere in the codebase:
   ```
   Unknown subcommand: {subcommand}
   Available subcommands: init, test, rewrite, git, build
   Run 'skim --help' for usage information.
   ```

2. **`resolve_invocation()` cyclomatic complexity** (Confidence: 82%) - The function has ~12 complexity branches. Consider extracting the file-like heuristic check into a named helper:
   ```rust
   fn looks_like_file_or_glob(token: &str) -> bool {
       token.contains('.') || token.contains('/') || token.contains('\\') 
           || token == "-" || token.contains('*') || token.contains('?') 
           || token.contains('[')
   }
   ```
   This improves readability and maintainability.

3. **Unused `SubcommandHandler` trait** (Confidence: 82%) - The trait is defined for Phase B extensibility but has no implementations and sits behind `#[allow(dead_code)]`. Per YAGNI, consider removing it and redesigning when Phase B work begins. Currently it adds cognitive overhead with no immediate value.

4. **`resolve_invocation()" reads `std::env::args()` directly** (Confidence: 85%) - Makes the function untestable in unit tests (only integration tests work). Consider extracting the core logic into a function that accepts `&[String]`:
   ```rust
   fn resolve_invocation_from_args(args: &[String]) -> Invocation { /* ... */ }
   fn resolve_invocation() -> Invocation {
       let raw_args: Vec<String> = std::env::args().collect();
       resolve_invocation_from_args(&raw_args[1..])
   }
   ```
   This enables fast unit tests for edge cases without process spawning.

**Test Coverage Notes:**

- Performance tests approve with no changes needed (9/10 score).
- Security review approves with no blocking issues found (9/10 score).
- Backward compatibility is well-managed by the conservative disambiguation logic.
- Main test improvement needed: Fix the two critical test gaps (already flagged as inline comments above).

### Recommendation
**CHANGES_REQUESTED** - Fix the 5 blocking issues before merge:
1. Replace `std::process::exit()` with proper return type handling
2. Add test assertion for `is_flag_with_value()` sync
3. Make `dispatch()` error handling consistent with codebase patterns
4. Add test for CWD-relative bare file named as subcommand
5. Add test for CWD-relative bare directory named as subcommand

Lower-confidence suggestions (complexity, error messages, trait removal) can be addressed in follow-up PRs or Phase B work.
